certbot-dns-google/certbot_dns_google/_internal/dns_google.py

```diff
@@ -188,7 +188,7 @@ def del_txt_record(self, domain, record_name, record_content, record_ttl):
 
         record_contents = self.get_existing_txt_rrset(zone_id, record_name)
         if record_contents is None:
-            record_contents = ["\"" + record_content + "\""]
+            record_contents = {"rrdatas": ["\"" + record_content + "\""], "ttl": record_ttl}
```

Comment:
there's something fishy about this whole block of code within the `del_txt_record` function. Why do we need to add `record_content` if `get_existing_txt_rrset` gave back `None` ? Am I missing some edge case here ? If the Plugin cycle goes like: `_setup_credentials -> _perform -> _cleanup`, then we don't need to cleanup what it is not dirty (i.e.: cleanup a record that we never created in the first place if for example `_perform` failed). What are your thoughts on this @alexzorin ?

Reply:
I haven't had time to look this over closely yet, but it appears to be related to https://github.com/certbot/certbot/pull/5678.

The required permissions of the plugin were expanded to include `.list`, in order to make wildcard certificates work. However,  the change was integrated in such a way that even if the `.list` permission was missing, the plugin would still try to delete the record optimistically, without knowing for sure that the record exists.

[See also the 0.22 release notes](https://github.com/certbot/certbot/blob/master/certbot/CHANGELOG.md#0220---2018-03-07).

Reply:
Hi Alex, I was able to reproduce this bug. I wanted to create a failing unit test but it's not possible to reproduce this way because we would need the whole plugin flow to run: `_setup_credentials -> _perform -> _cleanup`. So an easy way is to just apply the following patch ([gist](https://gist.github.com/lfundaro/308b00bcf8b5b637a0f3518b978e5eda)) and run certbot.

```
diff --git a/certbot-dns-google/certbot_dns_google/_internal/dns_google.py b/certbot-dns-google/certbot_dns_google/_internal/dns_google.py
index 22648fa37..f750ac295 100644
--- a/certbot-dns-google/certbot_dns_google/_internal/dns_google.py
+++ b/certbot-dns-google/certbot_dns_google/_internal/dns_google.py
@@ -156,6 +156,7 @@ class _GoogleClient(object):

         try:
             request = changes.create(project=self.project_id, managedZone=zone_id, body=data)
+            raise googleapiclient_errors.Error()
             response = request.execute()

             status = response['status']
```

As it should be expected from the Google API, you would get an HTTP 404 Not Found while trying to delete a record that doesn't exist:

```
googleapiclient.errors.HttpError: 
<HttpError 404 when requesting 
https://dns.googleapis.com/dns/v1/projects/foo-project/managedZones/12345/changes?alt=json 
returned "The 'entity.change.deletions[0]' resource named '_acme-challenge.non-existent-record.com. 
(TXT)' does not exist.">
```